File: Trie.js
class Node {
  constructor(character) {
    this.character = character;
    this.childrenHashTable = {}; //using hashtable instead of array
    this.isEndOfWord = false;
  }
  child = (char) => {
    return this.childrenHashTable[char];
  };
  putChild = (char) => {
    this.childrenHashTable[char] = new Node(char);
  };
  hasChild = (char) => {
    return this.childrenHashTable[char];
  };
  getChildren = () => {
    let result = [];
    for (let char in this.childrenHashTable) {
      result.push(this.childrenHashTable[char]);
    }
    return result;
  };
}

class Trie {
  constructor() {
    this.root = new Node(" ");
  }

  insert = (word) => {
    let current = this.root;
    for (let char of word) {
      if (!current.hasChild(char)) {
        current.putChild(char);
      }
      current = current.child(char);
    }
    current.isEndOfWord = true;
  };

  printTrie = () => {
    console.log(this.root);
  };

  contains = (word) => {
    if (word == null) return false;
    let current = this.root;
    for (let char of word) {
      if (!current.hasChild(char)) {
        return false;
      }
      current = current.child(char);
    }
    return current.isEndOfWord;
  };

  traversePreOrder = (currentNode) => {
    console.log(currentNode.character);

    for (let child of currentNode.getChildren()) {
      this.traversePreOrder(child);
    }
  };

  autoComplete = (prefix) => {
    let autoCompleteArray = [];
    let lastNode = this.findLastNode(prefix);
    this.findWord(lastNode, prefix, autoCompleteArray);
    return autoCompleteArray;
  };

  findWord = (currentNode, prefix, autoCompleteArray) => {
    if (currentNode.isEndOfWord) {
      autoCompleteArray.push(prefix);
    }
    for (let child of currentNode.getChildren()) {
      this.findWord(child, prefix + child.character, autoCompleteArray);
    }
  };

  findLastNode = (prefix) => {
    let current = this.root;
    for (let char of prefix) {
      if (!current.hasChild(char)) {
        return null;
      }
      current = current.child(char);
    }
    return current;
  };
}

let trie = new Trie();
trie.insert("canada");
trie.insert("car");
trie.insert("catch");
trie.insert("candy");
console.log(trie.autoComplete("can"));
// trie.traversePreOrder(trie.root);
// trie.printTrie();
// console.log(trie.contains("canada"));

File: Trie.py
class Node:
    def __init__(self, character):
        self.character = character
        self.childrenHashTable = {} #using hashtable instead of array
        self.isEndOfWord = False

    def __repr__(self):
        reprResult = "Node {}".format(self.character)
        return reprResult
    

    def child(self, char):
        return self.childrenHashTable.get(char)

    def putChild(self, char):
        self.childrenHashTable.update({char: Node(char)})

    def hasChild(self, char):
        return self.childrenHashTable.get(char)

    def getChildren(self):
        result = []
        for child in self.childrenHashTable:
            result.append(self.childrenHashTable[child])
        return result

class Trie:
    def __init__(self):
        self.root = Node(" ")

    def __repr__(self):
        return "Trie"
    
    def insert(self, word):
        current = self.root
        for char in word:
            if not current.hasChild(char):
                current.putChild(char)
            current = current.child(char)
        current.isEndOfWord = True

    def printTrie(self):
        print(self.root)


    def contains(self, word):
        if word == None:
            return False
        current = self.root
        for char in word:
            if not current.hasChild(char):
                return False
            current = current.child(char)
        return current.isEndOfWord


    def traversePreOrder(self, currentNode):
        print(currentNode.character)
        for child in currentNode.getChildren():
            self.traversePreOrder(child)

    def autoComplete(self, prefix):
        autoCompleteArray = []
        lastNode = self.findLastNode(prefix)
        self.findWord(lastNode, prefix, autoCompleteArray)
        return autoCompleteArray

    def findLastNode(self, prefix):
        current = self.root
        for char in prefix:
            if not current.hasChild(char):
                return None
            current = current.child(char)
        return current

    def findWord(self, currentNode, prefix, autoCompleteArray):
        if currentNode.isEndOfWord:
            autoCompleteArray.append(prefix)
        
        for child in currentNode.getChildren():
            self.findWord(child, prefix + child.character, autoCompleteArray)

trie = Trie()

trie.insert("canada")
trie.insert("candy")
trie.insert("car")
trie.insert("catch")
trie.insert("card")
print(trie.autoComplete("ca"))
# trie.traversePreOrder(trie.root)
#print(trie.contains("can"))